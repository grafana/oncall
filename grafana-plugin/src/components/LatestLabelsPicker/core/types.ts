export interface ItemRepresentation {
  id: string;
  repr: string;
}

export interface ItemGroup {
  key: ItemRepresentation;
  values: ItemRepresentation[];
}

export interface ItemSelected {
  key: ItemRepresentation;
  value: ItemRepresentation;
}