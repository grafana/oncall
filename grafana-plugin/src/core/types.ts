export interface ItemGroup {
  key: any;
  values: any[];
}

export interface ItemSelected {
  key: any;
  value: any;
}

export enum LabelInputType {
  key = 'key',
  value = 'value',
}

export interface ServiceLabelValidator {
  isValid: boolean;
  errorMessage?: string;
}
