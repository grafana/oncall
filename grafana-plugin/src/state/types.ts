import { AppPluginMeta, KeyValue } from '@grafana/data';

import { RootStore } from 'state/rootStore';

export interface WithStoreProps {
  store: RootStore;
}

export interface PageProps<T extends KeyValue = KeyValue> {
  meta: AppPluginMeta<T>;
  query: KeyValue;
}

export interface SelectOption {
  value: string | number;
  display_name: string;
}
