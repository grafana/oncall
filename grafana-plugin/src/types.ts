import { AppRootProps as BaseAppRootProps, AppPluginMeta, CurrentUserDTO, PluginConfigPageProps } from '@grafana/data';

export type OnCallPluginMetaJSONData = {
  stackId: number;
  orgId: number;
  onCallApiUrl: string;
  insightsDatasource?: string;
  license: string;
};

export type OnCallPluginMetaSecureJSONData = {
  grafanaToken: string;
  onCallApiToken: string;
};

export type AppRootProps = BaseAppRootProps<OnCallPluginMetaJSONData>;

// NOTE: it is possible that plugin.meta.jsonData is null (ex. on first-ever setup)
// the typing on AppPluginMeta does not seem correct atm..
export type OnCallAppPluginMeta = AppPluginMeta<null | OnCallPluginMetaJSONData>;
export type OnCallPluginConfigPageProps = PluginConfigPageProps<OnCallAppPluginMeta>;

// Extension points that other plugins can use to hook into the OnCall app.
export enum OnCallPluginExtensionPoints {
  AlertGroupAction = 'plugins/grafana-oncall-app/alert-group/action',
}

declare global {
  export interface Window {
    // https://github.com/grafana/grafana/blob/78bef7a26a799209b5307d6bde8e25fcb4fbde7d/public/views/index-template.html#L251-L258
    grafanaBootData: {
      user: CurrentUserDTO;
      settings: {
        unifiedAlertingEnabled: boolean;
        unifiedAlerting: { minInterval: string };
      };
    };
    RECAPTCHA_SITE_KEY: string;
    grecaptcha: any;
    dataLayer: any;
    mixpanel: any;
  }
}
