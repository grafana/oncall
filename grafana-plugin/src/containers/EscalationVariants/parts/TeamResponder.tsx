import React, { FC } from 'react';

import { HorizontalGroup, Select, IconButton } from '@grafana/ui';
import cn from 'classnames/bind';

import Avatar from 'components/Avatar/Avatar';
import PluginLink from 'components/PluginLink/PluginLink';
import Text from 'components/Text/Text';
import styles from 'containers/EscalationVariants/EscalationVariants.module.scss';
import {
  ResponderBaseProps,
  TeamResponder as TeamResponderType,
} from 'containers/EscalationVariants/EscalationVariants.types';

const cx = cn.bind(styles);

type TeamResponderProps = ResponderBaseProps & Pick<TeamResponderType, 'important' | 'data'>;

const TeamResponder: FC<TeamResponderProps> = ({ important, data, onImportantChange, handleDelete }) => (
  <li>
    <HorizontalGroup justify="space-between">
      <HorizontalGroup>
        <div className={cx('timeline-icon-background')}>
          <Avatar size="medium" src={data?.avatar_url} />
        </div>
        <Text className={cx('responder-name')}>{data.name}</Text>
        <Text type="secondary">by</Text>
        <Select
          className={cx('select')}
          width="auto"
          isSearchable={false}
          value={Number(important)}
          options={[
            {
              value: 0,
              label: 'Default',
              description: 'Use "Default notifications" from users personal settings',
            },
            {
              value: 1,
              label: 'Important',
              description: 'Use "Important notifications" from users personal settings',
            },
          ]}
          onChange={onImportantChange}
        />
        <Text type="secondary">notification policies</Text>
      </HorizontalGroup>
      <HorizontalGroup>
        <PluginLink className={cx('hover-button')} target="_blank" query={{ page: 'schedules', id: data.id }}>
          <IconButton
            tooltip="Open schedule in new tab"
            style={{ color: 'var(--always-gray)' }}
            name="external-link-alt"
          />
        </PluginLink>
        <IconButton className={cx('hover-button')} tooltip="Remove responder" name="trash-alt" onClick={handleDelete} />
      </HorizontalGroup>
    </HorizontalGroup>
  </li>
);

export default TeamResponder;
