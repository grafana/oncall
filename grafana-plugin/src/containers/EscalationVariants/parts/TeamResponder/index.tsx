import React, { FC } from 'react';

import { HorizontalGroup, IconButton } from '@grafana/ui';
import cn from 'classnames/bind';

import Avatar from 'components/Avatar/Avatar';
import Text from 'components/Text/Text';
import styles from 'containers/EscalationVariants/EscalationVariants.module.scss';
import {
  ResponderBaseProps,
  TeamResponder as TeamResponderType,
} from 'containers/EscalationVariants/EscalationVariants.types';
import NotificationPoliciesSelect from 'containers/EscalationVariants/parts/NotificationPoliciesSelect';

const cx = cn.bind(styles);

type TeamResponderProps = ResponderBaseProps & Pick<TeamResponderType, 'important' | 'data'>;

const TeamResponder: FC<TeamResponderProps> = ({
  important,
  data: { avatar_url, name },
  onImportantChange,
  handleDelete,
}) => (
  <li>
    <HorizontalGroup justify="space-between">
      <HorizontalGroup>
        <div className={cx('timeline-icon-background')}>
          <Avatar size="medium" src={avatar_url} />
        </div>
        <Text className={cx('responder-name')}>{name}</Text>
      </HorizontalGroup>
      <HorizontalGroup>
        <NotificationPoliciesSelect important={important} onChange={onImportantChange} />
        <IconButton
          data-testid="team-responder-delete-icon"
          tooltip="Remove responder"
          name="trash-alt"
          onClick={handleDelete}
        />
      </HorizontalGroup>
    </HorizontalGroup>
  </li>
);

export default TeamResponder;
