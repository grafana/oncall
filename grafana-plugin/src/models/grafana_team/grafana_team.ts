import { action, observable, makeObservable, runInAction } from 'mobx';

import { BaseStore } from 'models/base_store';
import { GrafanaTeam } from 'models/grafana_team/grafana_team.types';
import { makeRequest } from 'network/network';
import { RootStore } from 'state/rootStore';

type TeamItems = { [id: string]: GrafanaTeam };

export class GrafanaTeamStore extends BaseStore {
  @observable
  searchResult: Array<GrafanaTeam['id']> = [];

  @observable.shallow
  items: TeamItems = {};

  constructor(rootStore: RootStore) {
    super(rootStore);

    makeObservable(this);

    this.path = '/teams/';
  }

  @action.bound
  async updateTeam(id: GrafanaTeam['id'], data: Partial<GrafanaTeam>) {
    const result = await this.update(id, data);

    runInAction(() => {
      this.items = {
        ...this.items,
        [id]: result,
      };
    });
  }

  @action.bound
  async updateItems(query = '', includeNoTeam = true, onlyIncludeNotifiableTeams = false, short = true) {
    const result = await makeRequest<GrafanaTeam[]>(`${this.path}`, {
      params: {
        search: query,
        short: short ? 'true' : 'false',
        include_no_team: includeNoTeam ? 'true' : 'false',
        only_include_notifiable_teams: onlyIncludeNotifiableTeams ? 'true' : 'false',
      },
    });

    runInAction(() => {
      this.items = {
        ...this.items,
        ...result.reduce<TeamItems>(
          (acc, item) => ({
            ...acc,
            [item.id]: item,
          }),
          {}
        ),
      };

      this.searchResult = result.map((item: GrafanaTeam) => item.id);
    });
  }

  @action.bound
  async fetchItemById(id: string) {
    const team = await this.getById(id);

    runInAction(() => {
      this.items = {
        ...this.items,
        [id]: team,
      };
    });
  }

  getSearchResult = () => {
    return this.searchResult.map((teamId: GrafanaTeam['id']) => this.items[teamId]);
  };
}
