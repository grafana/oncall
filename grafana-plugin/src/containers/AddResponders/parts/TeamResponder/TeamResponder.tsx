import React, { FC } from 'react';

import { HorizontalGroup, IconButton } from '@grafana/ui';
import cn from 'classnames/bind';

import { Avatar } from 'components/Avatar/Avatar';
import { Text } from 'components/Text/Text';
import styles from 'containers/AddResponders/AddResponders.module.scss';
import { GrafanaTeam } from 'models/grafana_team/grafana_team.types';

const cx = cn.bind(styles);

type Props = {
  team: GrafanaTeam | null;
  handleDelete: React.MouseEventHandler<HTMLButtonElement>;
};

export const TeamResponder: FC<Props> = ({ team: { avatar_url, name }, handleDelete }) => (
  <li>
    <HorizontalGroup justify="space-between">
      <HorizontalGroup>
        <div className={cx('timeline-icon-background')}>
          <Avatar size="medium" src={avatar_url} />
        </div>
        <Text className={cx('responder-name')}>{name}</Text>
      </HorizontalGroup>
      <IconButton
        data-testid="team-responder-delete-icon"
        tooltip="Remove responder"
        name="trash-alt"
        onClick={handleDelete}
      />
    </HorizontalGroup>
  </li>
);
