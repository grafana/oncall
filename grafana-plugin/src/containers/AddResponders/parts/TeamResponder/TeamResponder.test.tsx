import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { GrafanaTeam } from 'models/grafana_team/grafana_team.types';

import { TeamResponder } from './TeamResponder';

describe('TeamResponder', () => {
  const team = {
    avatar_url: 'https://example.com',
    name: 'my test team',
  } as GrafanaTeam;

  test('it renders data properly', () => {
    const component = render(<TeamResponder team={team} handleDelete={() => {}} />);
    expect(component.container).toMatchSnapshot();
  });

  test('it calls the delete callback', async () => {
    const handleDelete = jest.fn();

    render(<TeamResponder team={team} handleDelete={handleDelete} />);

    const deleteIcon = await screen.findByTestId('team-responder-delete-icon');
    await userEvent.click(deleteIcon);

    expect(handleDelete).toHaveBeenCalledTimes(1);
  });
});
