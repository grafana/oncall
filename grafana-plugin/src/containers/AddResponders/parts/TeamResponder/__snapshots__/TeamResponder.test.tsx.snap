// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamResponder it renders data properly 1`] = `
<div>
  <li>
    <div
      class="css-1mhys9y-horizontal-group"
      style="width: 100%; height: 100%;"
    >
      <div
        class="css-18qv8yz-layoutChildrenWrapper"
      >
        <div
          class="css-ffyaiw-horizontal-group"
          style="width: 100%; height: 100%;"
        >
          <div
            class="css-18qv8yz-layoutChildrenWrapper"
          >
            <div
              class="timeline-icon-background"
            >
              <img
                class="root avatarSize-medium"
                data-testid="test__avatar"
                src="https://example.com"
              />
            </div>
          </div>
          <div
            class="css-18qv8yz-layoutChildrenWrapper"
          >
            <span
              class="root text text--undefined text--medium responder-name"
            >
              my test team
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-18qv8yz-layoutChildrenWrapper"
      >
        <button
          aria-label="Remove responder"
          class="css-17584xm"
          data-testid="team-responder-delete-icon"
          tabindex="0"
          type="button"
        >
          <div
            class="css-1j2891d-Icon"
          />
        </button>
      </div>
    </div>
  </li>
</div>
`;
